### PR TITLE
feat(tools): add tool_search for deferred MCP tool loading

### PR DESCRIPTION
**Pull request by nikuquwona**
## Summary

- Add **tool_search** feature: when `tool_search.enabled: true`, MCP tool schemas are deferred from the LLM context and discoverable at runtime via a `tool_search` tool, reducing prompt token usage
- Implement `DeferredToolRegistry` with three query forms: `select:name1,name2` (exact), `+keyword` (name filter + ranking), and `keyword` (regex search across name + description)
- Add `DeferredToolFilterMiddleware` using LangChain's `wrap_model_call` API to filter deferred tool schemas from `model.bind_tools()` while keeping them available in `ToolNode` for execution
- Enable `tool_name_prefix` in MCP client to prevent cross-server tool name collisions

## Changes

| File | Description |
|------|-------------|
| `tools/builtins/tool_search.py` | `DeferredToolRegistry`, singleton management, `tool_search` LangChain tool |
| `agents/middlewares/deferred_tool_filter_middleware.py` | Middleware to hide deferred schemas from `bind_tools` |
| `config/tool_search_config.py` | `ToolSearchConfig` pydantic model |
| `config/app_config.py` | Integrate `tool_search` config field |
| `tools/tools.py` | Register MCP tools in deferred registry when enabled |
| `agents/lead_agent/prompt.py` | Inject `<available-deferred-tools>` section into system prompt |
| `agents/lead_agent/agent.py` | Add `DeferredToolFilterMiddleware` to middleware chain |
| `mcp/tools.py` | Enable `tool_name_prefix=True` |
| `config.example.yaml` | Add `tool_search` config section |
| `tests/test_tool_search.py` | tests |

## Test plan

- [x] tests covering: `ToolSearchConfig`, `DeferredToolRegistry` (search, ranking, edge cases), singleton management, `tool_search` tool output, prompt section generation, `DeferredToolFilterMiddleware`
- [x] Full test suite passes (`make test` — 514 passed)
- [x] Lint passes (`make lint`)
- [x] Verified via LangGraph logs: `Tool search active: 26 tools deferred`, MCP tools load and execute correctly
- [ ] Manual test: agent discovers and invokes deferred tools via `tool_search` at runtime

### CODE REVIEW

File: backend/packages/harness/deerflow/tools/tools.py [hunk 1]
```diff
@@ -58,27 +76,23 @@ def get_available_tools(
...
+                        registry = DeferredToolRegistry()
+                        for t in mcp_tools:
+                            registry.register(t)
+                        set_deferred_registry(registry)
+                        builtin_tools.append(tool_search_tool)
+                        logger.info(f"Tool search active: {len(mcp_tools)} tools deferred")
         except ImportError:
             logger.warning("MCP module not available. Install 'langchain-mcp-adapters' package to enable MCP tools.")
         except Exception as e:
             logger.error(f"Failed to get cached MCP tools: {e}")
```
[the remark this answers is not in the document]
Comment (nikuquwona) [reply backend/packages/harness/deerflow/tools/tools.py:99]: fixed with always resetting deferred registry.

File: backend/packages/harness/deerflow/tools/builtins/tool_search.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+@tool
+def tool_search(query: str) -> str:
+    """Fetches full schema definitions for deferred tools so they can be called.
+
+    Deferred tools appear by name in <available-deferred-tools> in the system
+    prompt. Until fetched, only the name is known — there is no parameter
+    schema, so the tool cannot be invoked. This tool takes a query, matches
+    it against the deferred tool list, and returns the matched tools' complete
+    definitions. Once a tool's schema appears in that result, it is callable.
+
```
[the remark this answers is not in the document]
Comment (nikuquwona) [reply backend/packages/harness/deerflow/tools/builtins/tool_search.py:143]: Deferred tools are still registered in ToolNode for execution routing — bind_tools is only for schema injection. When the LLM sees the tool schema in tool_search output, it can emit a matching tool_call which ToolNode routes to the correct tool. This is by design.

File: backend/packages/harness/deerflow/agents/middlewares/deferred_tool_filter_middleware.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        registry = get_deferred_registry()
+        if not registry:
+            return request
+
+        deferred_names = {e.name for e in registry.entries}
+        active_tools = [t for t in request.tools if getattr(t, "name", None) not in deferred_names]
+
+        if len(active_tools) < len(request.tools):
+            logger.debug(f"Filtered {len(request.tools) - len(active_tools)} deferred tool schema(s) from model binding")
```
[the remark this answers is not in the document]
Comment (nikuquwona) [reply backend/packages/harness/deerflow/agents/middlewares/deferred_tool_filter_middleware.py:42]: ToolNode routes tool_call by name from its own registry, independent of bind_tools. No activation mechanism is needed.

File: backend/tests/test_tool_search.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+        mock_keys = [
+            "deerflow.agents",
+            "deerflow.agents.middlewares",
+            "deerflow.agents.middlewares.deferred_tool_filter_middleware",
+        ]
+        for key in mock_keys:
+            if isinstance(sys.modules.get(key), MagicMock):
+                del sys.modules[key]
+
```
[the remark this answers is not in the document]
Comment (nikuquwona) [reply backend/tests/test_tool_search.py:299]: Removing them mid-session allows real imports to work without affecting cleanup.
